### PR TITLE
WIP - FenixVersion

### PR DESCRIPTION
**Pull request by Callek**
Hey Johan,

could use your first-pass review here... I didn't yet touch tests.

### CODE REVIEW

File: mozilla_version/fenix.py [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+        (
+            -beta\.(?P<beta_number>\d+)
+            |-rc\.(?P<release_candidate_number>\d+)
+        )?$""", re.VERBOSE)
+
+    _OPTIONAL_NUMBERS = BaseVersion._OPTIONAL_NUMBERS + (
+        'beta_number', 'release_candidate_number'
+    )
+
+    _ALL_NUMBERS = BaseVersion._ALL_NUMBERS + _OPTIONAL_NUMBERS
```
Comment (JohanLorenzo) [inline mozilla_version/fenix.py]: Good call on putting this line. I wish we could get rid of it. I haven't found a way, so far.

File: mozilla_version/fenix.py [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+
+    beta_number = attr.ib(type=int, converter=strictly_positive_int_or_none, default=None)
+    release_candidate_number = attr.ib(
+        type=int, converter=strictly_positive_int_or_none, default=None
+    )
+
+    @classmethod
+    def parse(cls, version_string):
+        """Construct an object representing a valid Fenix version number."""
+        return super(FenixVersion, cls).parse(version_string)
```
Comment (JohanLorenzo) [inline mozilla_version/fenix.py]: Nit: I don't think we need to re-define the parse method, since we just delegate the call to the parents' method.

File: mozilla_version/fenix.py [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+        if isinstance(other, str):
+            other = FenixVersion.parse(other)
+        elif not isinstance(other, FenixVersion):
+            raise ValueError('Cannot compare "{}", type not supported!'.format(other))
+
+        difference = super(FenixVersion, self)._compare(other)
+        if difference != 0:
+            return difference
+
+        # XXX: Fix before landing:
```
Comment (JohanLorenzo) [inline mozilla_version/fenix.py]: 👍 

FWIW, Gecko implements this logic at:

https://github.com/mozilla-releng/mozilla-version/blob/2180fe058bc3ada6fa8cbdd24cc847f27b418798/mozilla_version/gecko.py#L340-L348

**Comment by Callek** [comment]
Ready for review, please see the bump logic, and let me know if I should change anything for MVP here.

File: mozilla_version/test/test_fenix.py [hunk 4]
```diff
@@ -0,0 +1,308 @@
...
+
+import mozilla_version.gecko
+
+from mozilla_version.errors import PatternNotMatchedError, TooManyTypesError, NoVersionTypeError
+from mozilla_version.fenix import FenixVersion
+from mozilla_version.test import does_not_raise
+
+
+VALID_VERSIONS = {
+    '0.1': 'release',
```
[the remark this answers is not in the document]
Comment (Callek) [reply mozilla_version/test/test_fenix.py]: of note, `0.1` was from official tags, but I can drop it for sure.

[the remark this answers is not in the document]
Comment (JohanLorenzo) [reply mozilla_version/test/test_fenix.py]: Oh gosh, I thought they've always follow semver, but `0.1` is the only counter-example 😅

Some `GeckoVersion`s have a way to specify known edge cases: 

https://github.com/mozilla-releng/mozilla-version/blob/2180fe058bc3ada6fa8cbdd24cc847f27b418798/mozilla_version/gecko.py#L416-L429

It might be too big just capture a single case. I'd say: let's not support `0.1` for now and enhance `FenixVersion` to support this case later. For instance, `GeckoVersion` didn't support old version numbers until https://github.com/mozilla-releng/mozilla-version/pull/59

File: mozilla_version/test/test_fenix.py [hunk 5]
```diff
@@ -0,0 +1,307 @@
...
+    3,
+    3.0,
+    ('3', '0', '1'),
+    ['3', '0', '1'],
+    LooseVersion('3.0'),
+    StrictVersion('3.0'),
+))
+def test_fenix_version_raises_eq_operator(wrong_type):
+    with pytest.raises(ValueError):
+        assert FenixVersion.parse('3.0') == wrong_type
```
Comment (JohanLorenzo) [inline mozilla_version/test/test_fenix.py]: Nit: `s/3.0/3.0.0` to get to 100% coverage https://coveralls.io/builds/27610891/source?filename=mozilla_version%2Ffenix.py#L98

File: mozilla_version/test/test_fenix.py [hunk 6]
```diff
@@ -0,0 +1,307 @@
...
+    # AttributeError is raised by LooseVersion and StrictVersion
+    with pytest.raises((ValueError, AttributeError)):
+        assert wrong_type == FenixVersion.parse('3.0')
```
Comment (JohanLorenzo) [inline mozilla_version/test/test_fenix.py]: Nit: `s/3.0/3.0.0`

File: mozilla_version/test/test_fenix.py [hunk 7]
```diff
@@ -0,0 +1,307 @@
...
+            major_number=major_number,
+            minor_number=minor_number,
+            patch_number=patch_number,
+            beta_number=beta_number,
+            release_candidate_number=rc_number,
+        )
+
+
+def test_fenix_version_constructor_minimum_kwargs():
+    assert str(FenixVersion(3, 0)) == '3.0'
```
Comment (JohanLorenzo) [inline mozilla_version/test/test_fenix.py]: Nit: Make `FenixVersion` throw an error whenever the patch number is missing.

File: mozilla_version/test/test_fenix.py [hunk 8]
```diff
@@ -0,0 +1,307 @@
...
+    assert str(FenixVersion(3, 0, 1)) == '3.0.1'
+    assert str(FenixVersion(3, 1, 0)) == '3.1.0'
+    assert str(FenixVersion(3, 0, beta_number=1)) == '3.0-beta.1'
+
+    assert str(FenixVersion(1, 0, release_candidate_number=1)) == '1.0-rc.1'
+
+
+@pytest.mark.parametrize('version_string, ExpectedErrorType', (
+    ('1.0.0b1', PatternNotMatchedError),
+    ('1.0.0.0b1', ValueError),
```
Comment (JohanLorenzo) [inline mozilla_version/test/test_fenix.py]: Question: I've been wondering if throwing 2 different type of errors (`PatternNotMatchedError` and `ValueError`) makes sense. Do you think `parse()` should throw only a single error, one day? 

(No action needed in this PR)

Comment (Callek) [reply mozilla_version/test/test_fenix.py]: Hrm, this is an interesting question ... not sure

Comment (JohanLorenzo) [reply mozilla_version/test/test_fenix.py]: No problem. We'll figure this out along the way.
